lib/queue_classic.rb

```diff
@@ -53,13 +53,17 @@ def self.has_connection?
   def self.default_conn_adapter
     t = Thread.current
     return t[:qc_conn_adapter] if t[:qc_conn_adapter]
-    adapter = if rails_connection_sharing_enabled?
-      ConnAdapter.new(ActiveRecord::Base.connection.raw_connection)
-    else
-      ConnAdapter.new
+
+    if rails_connection_sharing_enabled?
+      # AR connection should never be memoized, as call to
+      # ActiveRecord::Base.clear_active_connections! in current thread
+      # will cause it to return to AR pool, and it will become shared
+      # between memoized value and pool. And Rails does
+      # clear_active_connections! after each request
+      return ConnAdapter.new(ActiveRecord::Base.connection.raw_connection)
```

Comment:
I think it makes sense to move this part to the top (as an early `return`), because if `rails_connection_sharing_enabled?` is `true`, then `t = Thread.current` etc lines are effectively dead.

Reply:
There's another `return` before this, in order for setter to work. If thread-local `:qc_conn_adapter` is set before with setter, it's returned regardless of `rails_connection_sharing_enabled?`.

This setter is a little odd (even in master), as it sets default_conn_adapter only for current thread. After this changes, this behavior is retained. I think such setters are better to be deprecated.

Reply:
I see. For a future reviewer, if we forget about setter, the proposed change in essence is:

```ruby
def self.default_conn_adapter
  if rails_connection_sharing_enabled?
    ConnAdapter.new(ActiveRecord::Base.connection.raw_connection)
  else
    Thread.current[:qc_conn_adapter] ||= ConnAdapter.new
  end
end
```